myhpom/models/document.py

```diff
@@ -152,6 +211,220 @@ def authorized_client_ip(self, ip_address):
+        help_text="The raw content of the most recent response from CloudFactory",
+    )
+
+    def __unicode__(self):
+        return unicode(
+            "%s (%s %s)"
+            % (self.run_id, self.status, self.processed_at or self.created_at or self.inserted_at)
+        )
+
+    def create_post_data(self):
+        """Return the post data that CloudFactory expects when creating a run.
+        see https://docs.google.com/document/d/1VHD3iVq2Ky_SblQScv9O7tlphv9QHllMpGvn-bkWM_8/edit
+        """
+        ad = self.document_url.advancedirective
+        unit_data = {
+            'full_name': ad.user.get_full_name(),
+            'state': ad.user.userdetails.state.name if ad.user.userdetails.state else None,
+            'pdf_url': self.document_host + self.document_url.url,
+            'date_signed': str(ad.valid_date),
+        }
+        # Only submit non-blank values, to ensure valid input. (CloudFactory accepts blank values
+        # on POST, even though those values are invalid, but does not accept missing keys.)
+        # (The User's state can be null in our models. Others are ensured by model validations.)
+
+        unit_post_data = {k: v for k, v in unit_data.items() if bool(v) is True}
+
+        data = {
+            "line_id": settings.CLOUDFACTORY_LINE_ID,
+            "callback_url": self.document_host + reverse('myhpom:cloudfactory_response'),
+            "units": [unit_post_data],
+        }
+        return data
+
+    def save_response_data(self, response_content):
+        """update the response_data field and the other fields that are extracted from it.
+        """
+        # we want to save the response_content to self.response_data no matter what.
+        self.response_content = response_content
+        self.save()
+
+        # now we try to unpack the response_content on the assumption that it is json.
+        try:
+            data = json.loads(response_content)  # throws an error if not json
+            if 'id' in data:
+                self.run_id = data['id']
+            if 'status' in data:
+                self.status = data['status']
+            if 'created_at' in data:
+                self.created_at = data['created_at']
+            if 'processed_at' in data:
+                self.processed_at = data['processed_at']
+        except ValueError:
+            self.status = self.STATUS_ERROR
+            self.save()
+            raise
+
+        self.save()
+
+    def output(self):
+        """
+        Returns dictionary of 'output' from run if is processed.
+
+        Otherwise if the run failed to process returns None or return non-JSON
+        return None.
+        """
+        if self.status != CloudFactoryDocumentRun.STATUS_PROCESSED:
+            return None
+
+        try:
+            # no need to try/catch since we know it succeeded
+            data = json.loads(self.response_content)
+            units = data['units']
+            return units[0]['output']
+        except ValueError:
```

Comment:
If there is no 'output' block in the data, this raises a KeyError. Should we add 
```
        except KeyError:
            return None
```

Reply:
However, see below – maybe a better solution is to audit the callback request data to ensure the presence of an 'output' key.